### PR TITLE
Fix text cursor or input can be outside of the text field

### LINKED ISSUE #91695: Multiline TextField with inputed whitespaces has incorrect layout in some cases
I want to figure out reason of #90058 and fix it.
While I was looking into that, I found a few more problems with whitespaces input.
#### 1. Multiline TextField with TextAlign.center and inputed whitespaces at the end does not align these whitespaces with the input**
```
TextField(
  maxLines: null,
  decoration: const InputDecoration(filled: true),
  textAlign: TextAlign.center,
)
```
1_incorrect
If we will type more at the end (not whitespace) it is aligned correctly:
1_correct
This behavior is the same behavior on native iOS, but on native Android current line of input in multiline text field (last line) is aligned with whitespaces at the end:
android
This behavior should be different and correspond platforms native behaviors. 
The reason of #90058 is similar. That behavior is the same behavior on native iOS, on native Android whitespaces will be shown.

#### 2. Multiline TextField with TextDirection.rtl and inputed whitespaces moves cursor outside of the TextField
```
TextField(
  maxLines: null,
  decoration: const InputDecoration(filled: true),
  textDirection: TextDirection.rtl,
)
```

#### 3. Multiline TextField with TextAlign.left and TextDirection.rtl, and inputed whitespaces moves input outside of the TextField
```
TextField(
  maxLines: null,
  decoration: const InputDecoration(filled: true),
  textAlign: TextAlign.left,
  textDirection: TextDirection.rtl,
)
```

### PR DESCRIPTION
**Pull request by smitsk**
Fixes #91695
This fixes parts 2 and 3, part 1 will be fixed in engine and related to this.

### CODE REVIEW

**Comment by goderbauer** [comment]
@smitsk Looks like this PR is missing a test and failing some existing tests. Can you take a look at that?

/cc @justinmc @LongCatIsLooong

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 1]
```diff
@@ -821,7 +829,7 @@ class TextPainter {
       final TextBox box = boxes.last;
       final double caretStart = box.start;
       final double dx = box.direction == TextDirection.rtl ? caretStart - caretPrototype.width : caretStart;
-      return Rect.fromLTRB(min(dx, _paragraph!.width), box.top, min(dx, _paragraph!.width), box.bottom);
+      return Rect.fromLTRB(dx.clamp(0, _paragraph!.width), box.top, dx.clamp(0, _paragraph!.width), box.bottom);
```
[the remark this answers is not in the document]
Comment (smitsk) [reply packages/flutter/lib/src/painting/text_painter.dart:832]: So I fixed it by adding clamping in method
https://github.com/flutter/flutter/blob/04156b66364b6be646432cc77487c108fc19f378/packages/flutter/lib/src/rendering/editable.dart#L1519
I think it should be limited similar to caret metrics computation.

[the remark this answers is not in the document]
Comment (smitsk) [reply packages/flutter/lib/src/painting/text_painter.dart:832]: So I fixed it by adding clamping in method
https://github.com/flutter/flutter/blob/04156b66364b6be646432cc77487c108fc19f378/packages/flutter/lib/src/rendering/editable.dart#L1519
I think it should be constrained, like in caret metrics computation (because both methods use `getRectsForRange`)

[the remark this answers is not in the document]
Comment (smitsk) [reply packages/flutter/lib/src/painting/text_painter.dart:832]: So I fixed it by adding clamping in method
https://github.com/flutter/flutter/blob/04156b66364b6be646432cc77487c108fc19f378/packages/flutter/lib/src/rendering/editable.dart#L1519
I think x coordinate in the result should be clamped as when compute caret metrics (because both methods use `getRectsForRange`)

File: packages/flutter/lib/src/rendering/editable.dart [hunk 2]
```diff
@@ -1502,7 +1502,10 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
 
   // We need to check the paint offset here because during animation, the start of
   // the text may position outside the visible region even when the text fits.
-  bool get _hasVisualOverflow => _maxScrollExtent > 0 || _paintOffset != Offset.zero;
+  bool get _hasVisualOverflow {
```
[the remark this answers is not in the document]
Comment (smitsk) [reply packages/flutter/lib/src/rendering/editable.dart:1505]: This was originally planned to fix the part 3 of #91695
But I have already suggested another solution for that: flutter/engine#29429 
Could you take a look at this? If it's ok, I will revert this getter.

**Review by justinmc — COMMENTED** [review]
What's the native behavior on Android and iOS for this?  On Mac it doesn't show the spaces, which is not consistent with this PR:

| Native Mac TextEdit | Flutter on Mac with this PR |
| --- | --- |
| <video src="https://user-images.githubusercontent.com/389558/139729369-606dac9a-d0c0-4d00-82b3-85d3d85773c1.mov" /> | <video src="https://user-images.githubusercontent.com/389558/139741057-ef9249fa-2d63-4fa4-9136-69c92aa6e162.mov" /> |

It's also a bit confusing that the text goes off-screen horizontally even though the field is `maxLines: null` and should grow vertically.

It looks like this is a step in the right direction though.

**Comment by smitsk** [comment]
@justinmc @LongCatIsLooong 
> On Mac it doesn't show the spaces, which is not consistent with this PR

Behavior that is shown in your comment it's without this PR (Probably you mean PR flutter/engine#29429).

So let's look at some examples:
1. Right aligned ltr text
```
TextField(
  controller: TextEditingController(text: 'input      '),
  decoration: const InputDecoration(filled: true),
  textDirection: TextDirection.ltr,
  textAlign: TextAlign.right,
  maxLines: null,
),
```


Whitespaces isn't shown at the end of input. For caret metrics we get value from engine `getRectsForRange`. But this value is greater than container width. To avoid caret moves outside the container value from engine is limited by width:
`min(dx, _paragraph!.width)`  on framework side here
https://github.com/flutter/flutter/blob/9350a149a698b5f2970cc133c93c12696b8157a4/packages/flutter/lib/src/painting/text_painter.dart#L782 and here https://github.com/flutter/flutter/blob/9350a149a698b5f2970cc133c93c12696b8157a4/packages/flutter/lib/src/painting/text_painter.dart#L824
If we change these lines to `return Rect.fromLTRB(dx, box.top, dx, box.bottom);` caret will be outside:



I explained this behavior for left aligned text here:
https://github.com/flutter/flutter/pull/91698#issuecomment-951472775

2. Left aligned rtl text (It's symmetric to right aligned ltr text in first example, isn't it?)
```
TextField(
  controller: TextEditingController(text: 'input                        '),
  decoration: const InputDecoration(filled: true),
  textDirection: TextDirection.rtl,
  textAlign: TextAlign.left,
  maxLines: null,
),
```
Before flutter/engine#29429 whitespaces at the end are show and it moves input outside. I checked this by running on Android and on iOS. It's described in part 3 of https://github.com/flutter/flutter/issues/91695#issue-1024236653.
I think whitespaces shouldn't be shown. It seems logical to have the same behavior for similar cases (Left aligned rtl text and right aligned ltr text in first example).
And as you rightly noted native iOS does the same.
So flutter/engine#29429 fixes this, but then value for caret metrics from engine `getRectsForRange` can be outside the text container (it's similar to previous example) That's also shown in your video https://github.com/flutter/flutter/pull/91698#pullrequestreview-794643729.



To avoid this we should limit value from engine not only on the right side but also on the left here:
https://github.com/flutter/flutter/blob/9350a149a698b5f2970cc133c93c12696b8157a4/packages/flutter/lib/src/painting/text_painter.dart#L782
This PR does it by adding clamping.
If you run this example with changes flutter/engine#29429 and this PR, you will get:

File: packages/flutter/lib/src/rendering/editable.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (smitsk) [reply packages/flutter/lib/src/rendering/editable.dart]: Yes, It's reverted.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 3]
```diff
@@ -1529,8 +1529,8 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
       final Offset start = Offset(0.0, preferredLineHeight) + caretOffset + paintOffset;
       return <TextSelectionPoint>[TextSelectionPoint(start, null)];
     } else {
-      final Offset start = Offset(boxes.first.start, boxes.first.bottom) + paintOffset;
-      final Offset end = Offset(boxes.last.end, boxes.last.bottom) + paintOffset;
+      final Offset start = Offset(boxes.first.start.clamp(0, _textPainter.size.width), boxes.first.bottom) + paintOffset;
+      final Offset end = Offset(boxes.last.end.clamp(0, _textPainter.size.width), boxes.last.bottom) + paintOffset;
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/rendering/editable.dart:1533]: Is this change needed for fixing part 2 or 3? It makes sense to me but I think there're other public APIs that are dependent on `TextPainter.getBoxesForSelection`, so if we change `getEndpointsForSelection` we might want to update those us well to keep the behavior consistent.

**Comment by skia-gold** [comment]
Gold has detected about 6 new digest(s) on patchset 13.
View them at https://flutter-gold.skia.org/cl/github/91698

File: packages/flutter/lib/src/rendering/editable.dart (hunk 3 above)
Comment (smitsk) [reply packages/flutter/lib/src/rendering/editable.dart:1533]: This change doesn't fix part 2 or 3 directly, but it's required for correct `TextSelectionToolbar` location.
I checked text selection painting and updated it. Now we should update golden tests images. I described in the https://github.com/flutter/flutter/pull/91698#discussion_r738764737 that `letterSpacing` can bring the letters closer, but it can also move caret/selection outside (by value from `getRectsForRange`).
I will also check other items that are depends on `getBoxesForSelection`.

**Review by LongCatIsLooong — APPROVED** [review]
LGTM. I've triaged the golden diffs and hopefully the check will pass momentarily.
